Algorithms/Dynamic-Programming/rod_cutting.cpp

```diff
@@ -0,0 +1,108 @@
+// Output Format : The length(s) of the piece(s) leading to maximum cost followed by the maximum cost. 
+// Approach Used : Dynamic programming
+//
+// The solution uses bottom to top approach for dynamic programming. It is clear that the maximum price 
+// of a piece of length i is the same as  maximum among total cost of (piece of length j + piece of length(i-j) )
+// such that j<=i. Now the advantage of dynamic programming is that instead of using cost of piece of length(i-j)
+// we use its pre-calculated value/stored value(here price array). Since we use a bottom to up approach it is ensured
+// that this value is always computed before it is required. Hence price[n] easily gives us the maximum achievable price.
+// Additionally we use two arrays length_now and length_next to keep track of the pieces used to reach the maximum cost in 
+// a chain-like structure.
+
+// Author : monsijb
+
+#include<iostream>
+#include<vector>
+using namespace std;
+
+
+
+void rcProblem(vector<int>arr)
+{
+	int n = arr.size();
+	
+	// Auxiliary array in which price[i] stores the maximum price in which 
+	// the rod of length i can be sold. Clearly price[0]=0 
+	int price[n+1];
+	int current_length;
+	price[0]=0;
+	
+	// Arrays to keep track of the pieces used at each stage
+	int length_now[n+1];
+	int length_next[n+1];
+	
+	//Initialization of the two arrays
+	for(int i=0; i<=n; i++)
+	{
+		length_now[i]=-1;
+		length_next[i]=-1;
+	}
+	
+	for(int i=1; i<=n ;i++)
+	{
+		// Current maximum achievable price for rod of length i
+		int current_max = INT_MIN;
+		
+		for(int j=0; j<i; j++)
+		{
+			int current_price = arr[j] + price[i-(j+1)];
+			if(current_price > current_max)
+			{
+				current_max = current_price;
+				current_length = j;
+			}
+		}
+		
+		// Value calculated is stored appropriately(maximum cost of piece having length i) 
+		price[i] = current_max;
+		
+		// Lengths of pieces used to get piece of length i is also stored
+		length_now[i]=current_length+1;
+		length_next[i] = i-(length_now[i]);
+	}
+
+	// Starting from length n we move backwards to get what pieces are used to get the maximum cost 
+	int position=n;
+	cout << "Piece(s) of length : ";
+	
+	// Continously moving to next piece(as per length_next array)till the required length is reached.
+	while(length_next[position]!=0)
+	{
+		cout << length_now[position] << " ";
+		position = length_next[position];
+	}
+	cout << length_now[position] << "\n";
+	cout << "Maximum achievable price: " << price[n] << "\n";
+
+}
+
+int main()
+{
+	int n,x;
+	cout << "Enter the length of the rod: ";
+	cin >> n;
+	vector<int>arr;
+	cout << "Enter the cost of pieces having length 1 to " << n << "(" << n <<  " integers):\n"; 
+	for(int i=0; i<n; i++)
+	{
+		cin >> x;
+		arr.push_back(x);
+	}
+	rcProblem(arr);
+	return 0;
+}
+
+
+// Time complexity:O(n^2)
+// In the worst case the two nested loops may run for altogether 1+2+3+...n=(n(n+1))/2. So clearly the time complexity is O(n^2).
```

Comment:
Won't this be the complexity in each case, why just the worst case?

Reply:
The order will remain same althrough